File: 13.roman-to-integer.js
/*
 * @lc app=leetcode id=13 lang=javascript
 *
 * [13] Roman to Integer
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
function romanToInt(s) {
  const chars = s.split("");

  let num = 0;
  let current;
  let peek;

  for (let i = 0; i < chars.length; i++) {
    current = chars[i];
    peek = chars[i + 1];

    switch (current) {
      case "I":
        switch (peek) {
          case "V":
            num += 4;
            i++;
            break;
          case "X":
            num += 9;
            i++;
            break;
          default:
            num += 1;
            break;
        }

        break;
      case "X":
        switch (peek) {
          case "L":
            num += 40;
            i++;
            break;
          case "C":
            num += 90;
            i++;
            break;
          default:
            num += 10;
            break;
        }

        break;
      case "C":
        switch (peek) {
          case "D":
            num += 400;
            i++;
            break;
          case "M":
            num += 900;
            i++;
            break;
          default:
            num += 100;
            break;
        }

        break;

      case "V":
        num += 5;
        break;
      case "L":
        num += 50;
        break;
      case "D":
        num += 500;
        break;
      case "M":
        num += 1000;
        break;
    }
  }

  return num;
}
// @lc code=end

File: 13.roman-to-integer.rs
/*
 * @lc app=leetcode id=13 lang=rust
 *
 * [13] Roman to Integer
 */

// @lc code=start
impl Solution {
    pub fn roman_to_int(s: String) -> i32 {
        let mut iter = s.chars().peekable();
        let mut num = 0;

        loop {
            match iter.next() {
                Some(c) => match c {
                    'I' => match iter.peek() {
                        Some('V') => {
                            num += 4;
                            iter.next();
                        }
                        Some('X') => {
                            num += 9;
                            iter.next();
                        }
                        _ => num += 1,
                    },
                    'X' => match iter.peek() {
                        Some('L') => {
                            num += 40;
                            iter.next();
                        }
                        Some('C') => {
                            num += 90;
                            iter.next();
                        }
                        _ => num += 10,
                    },
                    'C' => match iter.peek() {
                        Some('D') => {
                            num += 400;
                            iter.next();
                        }
                        Some('M') => {
                            num += 900;
                            iter.next();
                        }
                        _ => num += 100,
                    },
                    'V' => num += 5,
                    'L' => num += 50,
                    'D' => num += 500,
                    'M' => num += 1000,
                    _ => {}
                },
                None => break,
            }
        }

        num
    }
}
// @lc code=end
